File: app/services/dataService.ts
import { QuizItem, ReelItem } from "@app/types/content";

import quizzesMock from "@mock/quizzes.json";
import reelsMock from "@mock/reels.json";

type DataState<T> = {
  cache: T[] | null;
  lastFetched: number | null;
};

const QUIZ_STATE: DataState<QuizItem> = {
  cache: null,
  lastFetched: null,
};

const REEL_STATE: DataState<ReelItem> = {
  cache: null,
  lastFetched: null,
};

const MOCK_NETWORK_DELAY = 180;

const toQuizItems = (data: unknown): QuizItem[] => {
  return (data as QuizItem[]).map((item) => ({
    ...item,
    options: {
      A: item.options.A,
      B: item.options.B,
      C: item.options.C,
      D: item.options.D,
    },
  }));
};

const toReelItems = (data: unknown): ReelItem[] => {
  return data as ReelItem[];
};

const simulateNetwork = async () =>
  new Promise((resolve) => setTimeout(resolve, MOCK_NETWORK_DELAY));

export const dataService = {
  getQuizzes: async (): Promise<QuizItem[]> => {
    if (QUIZ_STATE.cache) {
      return QUIZ_STATE.cache;
    }
    await simulateNetwork();
    const parsed = toQuizItems(quizzesMock);
    QUIZ_STATE.cache = parsed;
    QUIZ_STATE.lastFetched = Date.now();
    return parsed;
  },
  getReels: async (): Promise<ReelItem[]> => {
    if (REEL_STATE.cache) {
      return REEL_STATE.cache;
    }
    await simulateNetwork();
    const parsed = toReelItems(reelsMock);
    REEL_STATE.cache = parsed;
    REEL_STATE.lastFetched = Date.now();
    return parsed;
  },
  refreshQuizzes: async (): Promise<QuizItem[]> => {
    await simulateNetwork();
    const parsed = toQuizItems(quizzesMock);
    QUIZ_STATE.cache = parsed;
    QUIZ_STATE.lastFetched = Date.now();
    return parsed;
  },
  refreshReels: async (): Promise<ReelItem[]> => {
    await simulateNetwork();
    const parsed = toReelItems(reelsMock);
    REEL_STATE.cache = parsed;
    REEL_STATE.lastFetched = Date.now();
    return parsed;
  },
};

File: lib/services/data_service.dart
import 'dart:convert';
import 'package:flutter/services.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import '../models/content.dart';

class DataService {
  static List<QuizItem>? _quizCache;
  static List<ReelItem>? _reelCache;

  Future<List<QuizItem>> getQuizzes() async {
    if (_quizCache != null) return _quizCache!;
    await Future.delayed(const Duration(milliseconds: 180));
    final String assetPath = kIsWeb ? 'assets/assets/quizzes.json' : 'assets/quizzes.json';
    try {
      final String jsonString = await rootBundle.loadString(assetPath);
      final List<dynamic> jsonList = json.decode(jsonString);
      _quizCache = jsonList.map((json) => QuizItem.fromJson(json)).toList();
      return _quizCache!;
    } catch (e) {
      // Fallback to try the other path
      final String fallbackPath = kIsWeb ? 'assets/quizzes.json' : 'assets/assets/quizzes.json';
      try {
        final String jsonString = await rootBundle.loadString(fallbackPath);
        final List<dynamic> jsonList = json.decode(jsonString);
        _quizCache = jsonList.map((json) => QuizItem.fromJson(json)).toList();
        return _quizCache!;
      } catch (_) {
        return [];
      }
    }
  }

  Future<List<ReelItem>> getReels() async {
    if (_reelCache != null) return _reelCache!;
    await Future.delayed(const Duration(milliseconds: 180));
    final String assetPath = kIsWeb ? 'assets/assets/reels.json' : 'assets/reels.json';
    try {
      final String jsonString = await rootBundle.loadString(assetPath);
      final List<dynamic> jsonList = json.decode(jsonString);
      _reelCache = jsonList.map((json) => ReelItem.fromJson(json)).toList();
      return _reelCache!;
    } catch (e) {
      // Fallback to try the other path
      final String fallbackPath = kIsWeb ? 'assets/reels.json' : 'assets/assets/reels.json';
      try {
        final String jsonString = await rootBundle.loadString(fallbackPath);
        final List<dynamic> jsonList = json.decode(jsonString);
        _reelCache = jsonList.map((json) => ReelItem.fromJson(json)).toList();
        return _reelCache!;
      } catch (_) {
        return [];
      }
    }
  }

  Future<List<QuizItem>> refreshQuizzes() async {
    await Future.delayed(const Duration(milliseconds: 180));
    final String assetPath = kIsWeb ? 'assets/assets/quizzes.json' : 'assets/quizzes.json';
    try {
      final String jsonString = await rootBundle.loadString(assetPath);
      final List<dynamic> jsonList = json.decode(jsonString);
      _quizCache = jsonList.map((json) => QuizItem.fromJson(json)).toList();
      return _quizCache!;
    } catch (e) {
      final String fallbackPath = kIsWeb ? 'assets/quizzes.json' : 'assets/assets/quizzes.json';
      try {
        final String jsonString = await rootBundle.loadString(fallbackPath);
        final List<dynamic> jsonList = json.decode(jsonString);
        _quizCache = jsonList.map((json) => QuizItem.fromJson(json)).toList();
        return _quizCache!;
      } catch (_) {
        return [];
      }
    }
  }

  Future<List<ReelItem>> refreshReels() async {
    await Future.delayed(const Duration(milliseconds: 180));
    final String assetPath = kIsWeb ? 'assets/assets/reels.json' : 'assets/reels.json';
    try {
      final String jsonString = await rootBundle.loadString(assetPath);
      final List<dynamic> jsonList = json.decode(jsonString);
      _reelCache = jsonList.map((json) => ReelItem.fromJson(json)).toList();
      return _reelCache!;
    } catch (e) {
      final String fallbackPath = kIsWeb ? 'assets/reels.json' : 'assets/assets/reels.json';
      try {
        final String jsonString = await rootBundle.loadString(fallbackPath);
        final List<dynamic> jsonList = json.decode(jsonString);
        _reelCache = jsonList.map((json) => ReelItem.fromJson(json)).toList();
        return _reelCache!;
      } catch (_) {
        return [];
      }
    }
  }
}

